### PR TITLE
Conformance test for Config API

### PR DESCRIPTION
**Pull request by shivamkm07**
Signed-off-by: shivam <shivamkm07@gmail.com>

# Description
Adds Conformance test for Config API testing Redis Configuration Store (both v6 and v7). It also fixes a bug in redis unsubscribe operation

Also currently `configuration.UpdateEvent` doesn't have a way to notify user of the operation performed (e.g. set, delete etc.). Added this as well as logic for handling delete key scenario in redis. 

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #2325 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: configuration/redis/redis.go [hunk 1]
```diff
@@ -265,32 +265,34 @@ func (r *ConfigurationStore) Get(ctx context.Context, req *configuration.GetRequ
 
 func (r *ConfigurationStore) Subscribe(ctx context.Context, req *configuration.SubscribeRequest, handler configuration.UpdateHandler) (string, error) {
 	subscribeID := uuid.New().String()
+	keyStopChanMap := make(map[string]chan struct{})
 	if len(req.Keys) == 0 {
 		// subscribe all keys
 		stop := make(chan struct{})
-		r.subscribeStopChanMap.Store(subscribeID, stop)
+		keyStopChanMap[keySpaceAny] = stop
```
Comment (pravinpushkar) [inline configuration/redis/redis.go:272]: Should we use synchronised map here also ?

File: configuration/redis/redis.go [hunk 2]
```diff
@@ -324,19 +327,30 @@ func (r *ConfigurationStore) handleSubscribedChange(ctx context.Context, req *co
...
-		Metadata: req.Metadata,
-		Keys:     []string{targetKey},
-	})
-	if err != nil {
-		r.logger.Errorf("get response from redis failed: %s", err)
-		return
+	operation := msg.Payload
+	var items map[string]*configuration.Item
+
+	if operation == "del" {
```
Comment (pravinpushkar) [inline configuration/redis/redis.go:333]: why we are checking for del operation? AFAIK config api does not support `del` ops.

Comment (shivamkm07) [reply configuration/redis/redis.go:333]: Config API doesn't support delete as well as add/update keys. But we expect notification after subscribing if a key is updated/deleted. Currently if a redis subscribed key is deleted, it errors out in redis as it tries to `get` the keyValue but it is deleted. So it needs separate handling

File: configuration/redis/redis.go (hunk 1 above)
Comment (shivamkm07) [reply configuration/redis/redis.go:272]: `keyStopChanMap` is unique for a given subscription ID, unlike `subscribeStopChanMap` which is a common map for all subscription ids. imo it should be okay for this to be normal map.

Comment (pravinpushkar) [reply configuration/redis/redis.go:272]: got it. make sense.

**Comment by shivamkm07** [comment]
@berndverst @ItalyPaleAle tagging for early reviews.

File: configuration/redis/redis.go [hunk 3]
```diff
@@ -243,6 +243,10 @@ func (r *ConfigurationStore) Get(ctx context.Context, req *configuration.GetRequ
 
 		redisValue, err := r.client.Get(ctx, redisKey).Result()
 		if err != nil {
+			if err == redis.Nil {
+				r.logger.Warnf("redis key %s does not exist, ignore it\n", redisKey)
```
Comment (pravinpushkar) [inline configuration/redis/redis.go:247]:
```suggestion
				r.logger.Warnf("redis key %s does not exist, ignoring it\n", redisKey)
```

File: tests/utils/configupdater/redis/redis.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	separator = "||"
+)
+
+type metadata struct {
+	Host     string
+	Password string
+}
+
+type ConfigUpdater struct {
+	client redis.UniversalClient
```
[the remark this answers is not in the document]
Comment (berndverst) [reply tests/utils/configupdater/redis/redis.go:28]: You should add ConfigUpdater as an interface, and then implement in both the v8 and v9 client files within the internal package.

**Review by berndverst — CHANGES_REQUESTED** [review]

File: tests/utils/configupdater/redis/redis.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (shivamkm07) [reply tests/utils/configupdater/redis/redis.go]: Resolved using internal RedisClient interface.

**Comment by berndverst** [comment]
Please run your conformance test locally with redis 7. You will see it failing.

Similar to the RedisClient interface, you need to create a ConfigUpdater interface here:
https://github.com/dapr/components-contrib/blob/master/internal/component/redis/redis.go#L57

And then you need to provide implementations in both the v8 and v9 client (which correspond to redis 6 and 7).

https://github.com/dapr/components-contrib/blob/master/internal/component/redis/v8client.go
https://github.com/dapr/components-contrib/blob/master/internal/component/redis/v9client.go

**Review by berndverst — APPROVED** [review]

**Comment by berndverst** [comment]
I ran the conformance tests locally with Redis 6 and Redis 7

**Pull request merged**